Emscripten wget fixes

juj opened the pull request:
Some fixes:

1. `emscripten_wget()` and `emscripten_async_wget()` always called `add/removeRunDependency()`, which is a bit silly since these are C APIs, so they are called after application runtime has already been initialized, so it's too late for the run dependencies. 
2. Attempting to simultaneously wget the same URL twice would throw an exception, because of the above got called with identical dependency IDs.
3. wgetting to a destination directory on the filesystem that does not exist would lead to an exception "file not found". This changes the semantics to create the destination directory tree instead.

Regarding 1), I'm not quite sure if it would be a good use case to have static data initializers be able to call emscripten_wget() to preload stuff before main() has been called? If that it something that would make sense, then perhaps the fix for 2) will need to be something different.

kripken commented on src/library_fs.js:
```diff
@@ -628,6 +628,19 @@ mergeInto(LibraryManager.library, {
       return FS.mknod(path, mode, 0);
     },
+    // Creates a whole directory tree chain if it doesn't yet exist
+    mkdir_tree: function(path, mode) {
```
for consistency with other function names, how about `mkdirTree`? Or `makeTree`?

juj replied:
Ok, using `mkdirTree`, I think that is more informative than `makeTree`.

kripken commented on src/library_fs.js:
```diff
@@ -628,6 +628,19 @@ mergeInto(LibraryManager.library, {
+    mkdir_tree: function(path, mode) {
+      var dirs = path.split('/');
+      var d = '';
+      for(var i = 0; i < dirs.length; ++i) {
```
space after `for`

juj replied:
Check.

kripken commented on src/library_fs.js:
```diff
@@ -1759,39 +1772,39 @@ mergeInto(LibraryManager.library, {
           }
           if (onload) onload();
-          removeRunDependency(dep);
+          if (!noRunDep) removeRunDependency(dep);
```
these checks would be nicer to read, I think if they were `if (dep) ..` (which is functionally identical here).

kripken commented:
lgtm with those comments addressed.

I don't think static initializers are an issue, they would happen after the runtime is ready, like main().

gouletr commented:
@juj We exported a C function that we call in `preRun`, which in turn calls `emscripten_async_wget` to preload content needed before the app can run.

Will this commit:
1) prevent us from using `emscripten_async_run` in this context
2) break the runtime dependency, since we rely on this to prevent the main from starting until it finished downloading all content

We are essentially simulating `--preload-file` except the files are URLs instead of being packaged by the `file_packager.py`. Will this still work?

Thanks for fixing the directory tree creation, this will make our code cleaner!

juj commented:
Ended up rewriting the patch in a slightly different (simpler) way, and dropped the assertion for `runtimeInitialized`, I think that's not quite needed. The old semantics of the function `getUniqueRunDependency()` puzzle me, I wonder if the `#if ASSERTIONS` check there has been some kind of typo, or if there has been some larger thought there. In any case, fixed `getUniqueRunDependency()` to always return a guaranteed unique run dependency. In this manner `emscripten_wget`s do add and remove run dependencies when running, but they will just get ignored instead.

@gouletr: The `preRun` wgets should not be affected (especially now with the new form for the patch)

gouletr commented:
@juj ok thanks, I can test it on my side when it gets in. Thanks!

kripken commented on src/preamble.js:
```diff
@@ -1822,13 +1822,11 @@ var runDependencyTracking = {};
-#if ASSERTIONS
   var orig = id;
   while (1) {
     if (!runDependencyTracking[id]) return id;
```
`runDependencyTracking` will only exist is `ASSERTIONS`.

Did you run into a bug where run deps with duplicate names were added, and this led to a bug? The ids should only be used for debugging (in `ASSERTIONS`). The actually important thing is just the `runDependencies` counter.

juj replied:
Ahhh, now I understand. Yeah, the failing assert also only existed inside `ASSERTIONS`, I went and fixed this in two locations. Reverted this change, adding the `asyncLoad` to call `getUniqueRunDependency()` is enough.

The reason why I found the semantics here odd is that reading a call site of function `getUniqueRunDependency()`, e.g. in https://github.com/kripken/emscripten/blob/master/src/library_fs.js#L1766, makes it read like it should always generate a guaranteed unique name for the dependency, except that it does so only in assertions mode and does nothing in release.  Any case, updated the PR now to restore the assertion guard.

kripken commented:
lgtm
